### PR TITLE
feat: Implement EigenLayer Rewards V2 distribution

### PR DESCRIPTION
**Pull request by ahmadkaouk**
### Summary

This PR implements the EigenLayer Rewards Distribution V2 model for DataHaven, replacing the previous merkle-root-based rewards registry approach with EigenLayer's native `OperatorDirectedRewardsSubmission` API. This enables direct integration with EigenLayer's RewardsCoordinator for validator rewards distribution.

### Motivation

EigenLayer's V2 rewards model provides several advantages:
- **Direct integration**: Uses EigenLayer's native `createOperatorDirectedOperatorSetRewardsSubmission` API
- **Per-operator rewards**: Distributes rewards proportionally to individual operators based on their earned points
- **Simplified architecture**: Removes the need for a separate RewardsRegistry contract
- **Better UX**: Operators receive rewards directly through EigenLayer's established claiming mechanism

### Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                       DataHaven Substrate                       │
├─────────────────────────────────────────────────────────────────┤
│  Era End                                                        │
│    │                                                            │
│    ▼                                                            │
│  external-validators-rewards pallet                             │
│    │ generate_era_rewards_utils()                               │
│    │   • Calculate individual points per validator              │
│    │   • Compute total inflation amount                         │
│    │                                                            │
│    ▼                                                            │
│  RewardsSubmissionAdapter (runtime_common)                      │
│    │ build() → points_to_rewards() → encode_rewards_calldata()  │
│    │                                                            │
│    ▼                                                            │
│  Snowbridge Outbound Queue                                      │
│    │ CallContract(ServiceManager.submitRewards(...))            │
└────│────────────────────────────────────────────────────────────┘
     │
     ▼ Cross-chain message via Snowbridge
┌─────────────────────────────────────────────────────────────────┐
│                         Ethereum                                │
├─────────────────────────────────────────────────────────────────┤
│  DataHavenServiceManager                                        │
│    │ submitRewards(OperatorDirectedRewardsSubmission)           │
│    │   • Approve wHAVE tokens to RewardsCoordinator             │
│    │                                                            │
│    ▼                                                            │
│  EigenLayer RewardsCoordinator                                  │
│    │ createOperatorDirectedOperatorSetRewardsSubmission()       │
│    │                                                            │
│    ▼                                                            │
│  Operators claim rewards via EigenLayer                         │
└─────────────────────────────────────────────────────────────────┘
```

#### Smart Contracts (`contracts/`)

**DataHavenServiceManager.sol**
- Added `submitRewards(OperatorDirectedRewardsSubmission)` function to submit rewards to EigenLayer's RewardsCoordinator
- Implements `SafeERC20` for secure token approvals
- Uses `onlyRewardsInitiator` modifier for access control (Snowbridge Agent)
- Emits `RewardsSubmitted` and `RewardsInitiatorSet` events for tracking

**IDataHavenServiceManager.sol**
- Added `submitRewards()` interface for EigenLayer rewards submission
- Added `setRewardsInitiator()` interface for configuring the authorized caller
- Added new events: `RewardsSubmitted`, `RewardsInitiatorSet`

**New Test: RewardsSubmitter.t.sol**
- Comprehensive test suite covering:
  - Access control (only rewards initiator can submit)
  - Single and multiple operator rewards
  - Multiple consecutive submissions
  - Custom descriptions and different tokens

#### Substrate Runtime (`operator/`)

**New: `runtime/common/src/rewards_adapter.rs` (934 lines)**

A generic, configurable adapter for building EigenLayer rewards messages:

- **`RewardsSubmissionConfig` trait**: Runtime-agnostic configuration interface
  - `OutboundQueue`: Snowbridge outbound queue type
  - `rewards_duration()`: Reward period duration (typically 86400s)
  - `whave_token_address()`: wHAVE ERC20 token on Ethereum
  - `service_manager_address()`: ServiceManager contract address
  - `rewards_agent_origin()`: Snowbridge agent origin

- **`RewardsSubmissionAdapter<C>`**: Generic implementation of `SendMessage` trait

- **`points_to_rewards()`**: Converts validator points to token amounts
  - Proportional distribution based on total points
  - Returns remainder (dust) from integer division
  - Arithmetic overflow/underflow protection

- **`encode_rewards_calldata()`**: ABI-encodes the `submitRewards` call
  - Uses `alloy-core` for type-safe Solidity ABI encoding
  - Validates `uint96` multiplier bounds

- **Comprehensive test suite** covering:
  - Basic and edge-case reward calculations
  - Remainder/dust handling
  - Overflow/underflow protection
  - ABI encoding round-trip verification
  - Message building with various configurations

**Modified: `pallets/external-validators-rewards/`**

- **`types.rs`**: Extended `EraRewardsUtils` struct:
  ```rust
  pub struct EraRewardsUtils {
      pub era_index: u32,                    // NEW
      pub rewards_merkle_root: H256,
      pub leaves: Vec<H256>,
      pub leaf_index: Option<u64>,
      pub total_points: u128,
      pub individual_points: Vec<(H160, u32)>, // NEW
      pub inflation_amount: u128,             // NEW
      pub era_start_timestamp: u32       // NEW
  }
  ```

- **`lib.rs`**: Updated `generate_era_rewards_utils()`:
  - Now accepts `inflation_amount` parameter
  - Extracts `individual_points` as `(H160, u32)` tuples for EigenLayer
  - Returns `None` when `total_points` is zero (prevents inflation with no distribution)

- **`mock.rs`**: Updated test mock to use `H160` as `AccountId` (matching DataHaven's EVM-compatible account model)

**Modified: Runtime Configurations**

All three runtimes (mainnet, stagenet, testnet) updated:

1. **New runtime parameters** (`runtime_params.rs`):
   - `ServiceManagerAddress`: DataHaven ServiceManager contract on Ethereum
   - `WHAVETokenAddress`: wHAVE ERC20 token address
   - `RewardsGenesisTimestamp`: EigenLayer-aligned genesis timestamp
   - `RewardsDuration`: Rewards period (default: 86400 = 1 day)

2. **Refactored `RewardsSendAdapter`**:
   - Replaced inline implementation with `RewardsSubmissionAdapter<Config>`
   - Each runtime implements `RewardsSubmissionConfig` trait
   - Cleaner, DRY configuration

## ⚠️ Breaking Changes ⚠️

- **Runtime Parameters**: New parameters must be configured via governance before rewards submission will work:
  - `ServiceManagerAddress` (replaces `RewardsRegistryAddress`)
  - `WHAVETokenAddress`
  - `RewardsGenesisTimestamp`
  
- **Contract Interface**: `submitRewards()` now accepts a full `OperatorDirectedRewardsSubmission` struct instead of a merkle root

### CODE REVIEW

File: operator/runtime/mainnet/src/configs/runtime_params.rs [hunk 1]
```diff
@@ -358,6 +358,28 @@ pub mod dynamic_params {
...
+        /// The wHAVE ERC20 token address on Ethereum.
+        /// Used in the OperatorDirectedRewardsSubmission struct.
+        pub static WHAVETokenAddress: H160 = H160::repeat_byte(0x0);
+
+        #[codec(index = 41)]
+        #[allow(non_upper_case_globals)]
+        /// EigenLayer-aligned genesis timestamp for rewards calculation.
+        /// Must be divisible by 86400 (seconds per day) as per EigenLayer requirements.
+        /// Default: 0 (must be set via governance to actual deployment timestamp).
+        pub static RewardsGenesisTimestamp: u32 = 0;
```
Comment (gonzamontiel) [inline operator/runtime/mainnet/src/configs/runtime_params.rs:406]: I believe this became obsolete since this commit https://github.com/datahaven-xyz/datahaven/pull/351/changes/b8c8ea30ef9a799632cda2bb002bc1b9c37a322f, when the calculation for `era_start_timestamp()` was moved inside the pallet's `on_era_end()`

File: operator/runtime/stagenet/src/configs/runtime_params.rs [hunk 2]
```diff
@@ -363,6 +363,28 @@ pub mod dynamic_params {
...
+        /// The wHAVE ERC20 token address on Ethereum.
+        /// Used in the OperatorDirectedRewardsSubmission struct.
+        pub static WHAVETokenAddress: H160 = H160::repeat_byte(0x0);
+
+        #[codec(index = 41)]
+        #[allow(non_upper_case_globals)]
+        /// EigenLayer-aligned genesis timestamp for rewards calculation.
+        /// Must be divisible by 86400 (seconds per day) as per EigenLayer requirements.
+        /// Default: 0 (must be set via governance to actual deployment timestamp).
+        pub static RewardsGenesisTimestamp: u32 = 0;
```
Comment (gonzamontiel) [inline operator/runtime/stagenet/src/configs/runtime_params.rs:409]: Same as prev comment

File: operator/runtime/testnet/src/configs/runtime_params.rs [hunk 3]
```diff
@@ -359,6 +359,25 @@ pub mod dynamic_params {
...
+
+        #[codec(index = 40)]
+        #[allow(non_upper_case_globals)]
+        /// The wHAVE ERC20 token address on Ethereum.
+        pub static WHAVETokenAddress: H160 = H160::repeat_byte(0x0);
+
+        #[codec(index = 41)]
+        #[allow(non_upper_case_globals)]
+        /// EigenLayer-aligned genesis timestamp for rewards calculation.
+        pub static RewardsGenesisTimestamp: u32 = 0;
```
Comment (gonzamontiel) [inline operator/runtime/testnet/src/configs/runtime_params.rs:404]: Same as prev comment

File: operator/runtime/testnet/src/configs/runtime_params.rs
Comment (gonzamontiel) [inline operator/runtime/testnet/src/configs/runtime_params.rs]: Can't we get rid of this parameter entirely now that we should be using the `DatahavenServiceManagerAddress`. Same for the other runtimes.

File: operator/pallets/external-validators-rewards/src/lib.rs [hunk 4]
```diff
@@ -308,7 +345,6 @@ pub mod pallet {
 
             // Mint tokens using the configurable handler
             let ethereum_sovereign_account = T::RewardsEthereumSovereignAccount::get();
-            let inflation_amount = T::EraInflationProvider::get();
             if let Err(err) =
                 T::HandleInflation::mint_inflation(&ethereum_sovereign_account, inflation_amount)
```
Comment (gonzamontiel) [inline operator/pallets/external-validators-rewards/src/lib.rs]: There's a possible corner/initial case in which the network is live, and some validators have non zero points, but the addresses are not configured properly yet (`WHAVETokenAddress` or `DatahavenServiceManagerAddress` are zero, for example).
In this case, `mint_inflation()` would still be called, but no message will make it to EigenLayer, so we would be indeed minting inflation with no-distribution. Maybe is unlikely because we need to configure those parameters before, but to be on the safe side we can also avoid minting if the addresses are mis-configured.

File: operator/runtime/common/src/rewards_adapter.rs [hunk 5]
```diff
@@ -0,0 +1,925 @@
...
+        return None;
+    }
+
+    if remainder > 0 {
+        log::debug!(target: LOG_TARGET, "Reward distribution remainder (dust): {} tokens", remainder);
+    }
+
+    let calldata = encode_rewards_calldata(
+        whave_token_address,
+        &C::strategies_and_multipliers(),
```
Comment (gonzamontiel) [inline operator/runtime/common/src/rewards_adapter.rs]: Right now this is an empty vec, but this can easily change and break if the implementation doesn't respect the commment:

> /// EigenLayer requires `strategiesAndMultipliers` to be sorted by strategy address (ascending)
> /// with no duplicates. Multipliers must fit in `uint96`.

Not mandatory, but we could have a validation before sending the rewards that this is indeed sorted, etc.

Comment (gonzamontiel) [reply operator/runtime/common/src/rewards_adapter.rs]: Or we can also add a check in `DataHavenServiceManagerr.sol`, in `submitRewards()`, at the very beginning before summing the amounts.

File: operator/runtime/common/src/rewards_adapter.rs [hunk 6]
```diff
@@ -0,0 +1,925 @@
...
+///
+/// # Arguments
+/// * `points` - List of (operator, points) tuples
+/// * `total_points` - Sum of all points
+/// * `inflation` - Total tokens to distribute
+///
+/// # Returns
+/// `Ok((operator_rewards, remainder))` where:
+/// * `operator_rewards` - Sorted list of (operator_address, amount) tuples
+/// * `remainder` - Dust amount from integer division truncation
```
Comment (gonzamontiel) [inline operator/runtime/common/src/rewards_adapter.rs:233]: Do we have a plan for this one? I know is very little, but over time maybe we want to do something since it's being minted? Just a thought for the future.

File: contracts/src/DataHavenServiceManager.sol [hunk 7]
```diff
@@ -232,6 +236,48 @@ contract DataHavenServiceManager is ServiceManagerBase, IDataHavenServiceManager
...
+            operatorSet, submissions
+        );
+
+        emit RewardsSubmitted(totalAmount, submission.operatorRewards.length);
+    }
+
+    /// @notice Sets the rewards initiator address (overrides deprecated base implementation)
+    /// @param newRewardsInitiator The new rewards initiator address
+    /// @dev Only callable by the owner
+    function setRewardsInitiator(
```
Comment (gonzamontiel) [inline contracts/src/DataHavenServiceManager.sol:271]: If for any reason this rewardsInitiator is misconfigured (or for example the zero address), every rewards submission tx from the bridge reverts with `OnlyRewardsInitiator()`. In this scenario `on_era_end` will still mint inflation and we will still mint gas from the failed `submitRewards()`. Maybe we should add an extra check here or in `submitRewards()`.

EDIT: this is not needed because we want to mint linearly (so on every on_era_end() )

File: contracts/test/RewardsSubmitter.t.sol [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        IRewardsCoordinatorTypes.OperatorDirectedRewardsSubmission memory submission =
+            _buildSubmission(rewardAmount, operator1);
+
+        // Warp to a time after the period ends
+        vm.warp(submission.startTimestamp + submission.duration + 1);
+
+        vm.prank(snowbridgeAgent);
+        vm.expectEmit(false, false, false, true);
+        emit IDataHavenServiceManagerEvents.RewardsSubmitted(rewardAmount, 1);
+        serviceManager.submitRewards(submission);
```
Comment (undercover-cactus) [inline contracts/test/RewardsSubmitter.t.sol:123]: I don't see the operator validator being registered in this test. I am wrong to think that a validator should be register as a validator to get a reward ?

File: operator/runtime/common/src/rewards_adapter.rs [hunk 9]
```diff
@@ -0,0 +1,925 @@
...
+pub enum RewardsAdapterError {
+    /// A strategy multiplier exceeds the maximum value for uint96.
+    InvalidMultiplier,
+    /// An arithmetic multiplication overflowed.
+    MultiplicationOverflow,
+    /// An arithmetic division by zero.
+    DivisionByZero,
+}
+
+sol! {
```
Comment (undercover-cactus) [inline operator/runtime/common/src/rewards_adapter.rs:57]: This is a really useful macro. It avoid writing painful encoding. Could you confirm it is working correctly in the runtime ? It doesn't rely on `std` ?

Comment (ahmadkaouk) [reply operator/runtime/common/src/rewards_adapter.rs:57]: Yes, it's working correctly and don't rely on `std`

File: operator/runtime/mainnet/src/configs/runtime_params.rs (hunk 1 above)
Comment (ahmadkaouk) [reply operator/runtime/mainnet/src/configs/runtime_params.rs:406]: This will be used in a follow-up PR about the rewards distribution.

File: operator/runtime/testnet/src/configs/runtime_params.rs
Comment (ahmadkaouk) [reply operator/runtime/testnet/src/configs/runtime_params.rs]: Fixed

File: operator/runtime/common/src/rewards_adapter.rs (hunk 5 above)
Comment (ahmadkaouk) [reply operator/runtime/common/src/rewards_adapter.rs]: Fixed in [f0e55c3](https://github.com/datahaven-xyz/datahaven/pull/351/commits/f0e55c30c3d5ba4f65852fce92aae4bdee487ab8)

File: operator/pallets/external-validators-rewards/src/lib.rs (hunk 4 above)
Comment (ahmadkaouk) [reply operator/pallets/external-validators-rewards/src/lib.rs]: Not really related to the changes in this PR, but as discussed, we should probably mint inflation consistently to ensure predictable inflation. On the other hand, to send the reward message, there is already a check of addresses. https://github.com/datahaven-xyz/datahaven/blob/d42a9283e2dd94e1c9719b7426be4ff93ad0c1ac/operator/runtime/common/src/rewards_adapter.rs#L150-L161

File: operator/runtime/common/src/rewards_adapter.rs (hunk 6 above)
Comment (ahmadkaouk) [reply operator/runtime/common/src/rewards_adapter.rs:233]: Fixed in [e824906](https://github.com/datahaven-xyz/datahaven/pull/351/commits/e8249062a7b12b5137757632fbc7aa295fbc3f50). Remainder is transferred to the treasury account.

File: operator/runtime/common/src/rewards_adapter.rs [hunk 10]
```diff
@@ -0,0 +1,929 @@
...
+use sp_std::vec::Vec;
+
+/// Default description for rewards submissions.
+pub const REWARDS_DESCRIPTION: &str = "DataHaven validator rewards";
+
+/// Log target for rewards adapter messages.
+const LOG_TARGET: &str = "rewards_adapter";
+
+/// Gas limit for the submitRewards call on Ethereum.
+pub const SUBMIT_REWARDS_GAS_LIMIT: u64 = 2_000_000;
```
Comment (undercover-cactus) [inline operator/runtime/common/src/rewards_adapter.rs:44]: How did you calculate the `SUBMIT_REWARDS_GAS_LIMIT` here ?

File: operator/runtime/common/src/rewards_adapter.rs [hunk 11]
```diff
@@ -0,0 +1,929 @@
...
+        gas: SUBMIT_REWARDS_GAS_LIMIT,
+        value: 0,
+    }]
+    .try_into()
+    .ok()?;
+
+    Some(OutboundMessage {
+        origin: C::rewards_agent_origin(),
+        id: H256::from_low_u64_be(rewards_utils.era_index as u64).into(),
+        fee: 0,
```
Comment (undercover-cactus) [inline operator/runtime/common/src/rewards_adapter.rs:214]: Why the fee here is 0 ?

File: operator/runtime/common/src/rewards_adapter.rs (hunk 10 above)
Comment (ahmadkaouk) [reply operator/runtime/common/src/rewards_adapter.rs:44]: I didn’t calculate the value. I simply passed a limit that is high enough to ensure the call has sufficient gas. If the gas limit is too low, the entire cross-chain message fails and cannot be retried.

File: operator/runtime/common/src/rewards_adapter.rs (hunk 11 above)
Comment (ahmadkaouk) [reply operator/runtime/common/src/rewards_adapter.rs:214]: fee currently has no practical impact since the relayers are configured to skip fee check. https://github.com/datahaven-xyz/datahaven/blob/42ec577f15fbd0e62eda12d22e2df479510def19/test/configs/snowbridge/stagenet/solochain-relay.json#L31-L35
This will change once we enable fee check.

**Review by gonzamontiel — APPROVED** [review]
Nice work. I'm fine with merging it !
